src/cargo/ops/cargo_compile.rs

```diff
@@ -203,7 +204,7 @@ pub fn compile_pkg<'a>(root_package: &Package,
         None => {
             for &to_build in to_builds.iter() {
                 let targets = try!(generate_targets(to_build, profiles, mode,
-                                                    filter, release));
+                                                    filter, all_feature, release));
```

Comment:
The list of all features should actually come from the features enabled for the `to_build` package rather than the root package in this case.